### ISSUE TITLE
JAX tree_* drops `None` elements

### ISSUE THREAD

**Issue by tomhennigan** [google/jax#2348]
Somewhat confusingly (imho) flatten drops leaves that are `None`:

```python
>>> jax.tree_leaves([1, None, 1])
[1, 1]
```

However flatten/unflatten on such a structure is symmetrical (leading me to believe JAX treats a `None` leaf as part of the structure):

```python
>>> leaves, treedef = jax.tree_flatten([1, None, 1])
>>> jax.tree_unflatten(treedef, leaves)
[1, None, 1]
```

However multimap does not complain that the structures are different, and also skips the map fn if the left hand leaf is `None`:

```python
>>> a = [None, 1]
>>> b = [2, None]
>>> jax.tree_multimap(lambda x, y: (x, y), a, b)
[None, (1, None)]
```

Is this desired behaviour? It doesn't align with the other tree libraries (e.g. tf.nest and dm-tree) which treat `None` as a leaf, so it feels broken to me as a user of those libraries.

**Comment by tomhennigan** [comment]
(I should add that not everyone I've spoken to agrees with my preference and some folks prefer the current "None is part of the structure" approach.)

**Comment by girving** [comment]
To expand on why I like the current behavior of treating `None` as structure and not leaf: typically when I'm writing utilities that work on pytrees the utilities will only work if the leaves all have a certain type `T`.  Thus, such utilities are more general if more of the pytree is viewed as structure rather than leaves, so that there are fewer leaves and thus more of them have type `T`.

The most important case of this is that the toplevel pytree `None` should be viewed as having no leaves, so that any utility which maps across pytrees will work if `None` is passed in.  For example, `jax.lax.scan` automatically optimally handles information threaded into each layer and out of each layer, since you can use `None` pytree in the appropriate place.

However, I agree that the `jax.tree_multimap` example @tomhennigan showed is odd, since I'd expect it to assert that the structures don't match.

**Comment by mattjj** [comment]
I wonder if a good option is to use dm-tree when you want dm-tree behavior, rather than using JAX pytrees for that particular use case. WDYT?

**Comment by tomhennigan** [comment]
Thank you for all the context.

I am still not convinced that this is the right choice for the user facing API, but I understand that this is deliberate, appreciate that it is documented and acknowledge that changing it now would be a significant amount of work.

I'll close this out because I suspect JAX team have many more important things to work on and think about.

BTW I think we have not seen this much since JAX itself typically does not return `None` (e.g. unconnected gradients are zeros not `None` unlike tf), but user code might reasonably do this and I think many folks are using `jax.tree_*` to avoid additional deps (on TF or dm-tree).

The workaround of using [dm-tree](https://github.com/deepmind/tree) in cases that require `None` as leaves is fine for us, although I suspect we'll see user confusion with JAX tree utils since many people are familiar with the semantics of `tf.nest`/`tree`.

**Comment by skye** [comment]
Just my $.02, I would prefer if this were behavior were switchable, as I also find the current behavior surprising and hard to remember (I screwed up using pytrees with Nones not five minutes ago!) and I don't like the vmap dance. But agreed this isn't top priority.

**Comment by alvarosg** [comment]
+1 of documenting this more thoroughly in the doc strings of methods in `tree_util`. I don't see any mentions to `None` [in the tree_utils doc](https://jax.readthedocs.io/en/latest/jax.tree_util.html)

And the [notebook](https://jax.readthedocs.io/en/latest/notebooks/JAX_pytrees.html) referenced earlier does not seem exist anymore?
